### ISSUE TITLE
Conflict/Integration with hbac.nvim

### ISSUE THREAD

**Issue by DrKGD** [Hajime-Suzuki/vuffers.nvim#44]
Hi there, me again, with a different topic this time.

Every once in a while I had a nvim error popping up at the worst possible times, I was not sure if it was either due to a plugin or to nvim itself (as I always use the nightly :yum:)

<details>
  <summary>Spoiler: error contents</summary>

```
E5108: Error executing lua: vim/_editor.lua:0: nvim_exec2()..BufEnter Autocommands for "*": Vim(append):Error executing lua callback: ...rte/.local/share/nvim/lazy/ui.vuffers/lua/vuffers/ui.lua:67: Invalid buffer id: 8
stack traceback:
	[C]: in function '__index'
	...rte/.local/share/nvim/lazy/ui.vuffers/lua/vuffers/ui.lua:67: in function 'f'
	...rte/.local/share/nvim/lazy/ui.vuffers/lua/utils/list.lua:138: in function 'map'
	...rte/.local/share/nvim/lazy/ui.vuffers/lua/vuffers/ui.lua:241: in function 'handler'
	...cal/share/nvim/lazy/ui.vuffers/lua/vuffers/event-bus.lua:53: in function 'callback'
	...rte/.local/share/nvim/lazy/ui.vuffers/lua/utils/list.lua:19: in function 'for_each'
	...cal/share/nvim/lazy/ui.vuffers/lua/vuffers/event-bus.lua:51: in function 'publish'
	...cal/share/nvim/lazy/ui.vuffers/lua/vuffers/event-bus.lua:77: in function 'publish_buffer_list_changed'
	.../share/nvim/lazy/ui.vuffers/lua/vuffers/buffers/init.lua:116: in function 'set_active_buf'
	...share/nvim/lazy/ui.vuffers/lua/vuffers/auto-commands.lua:37: in function <...share/nvim/lazy/ui.vuffers/lua/vuffers/auto-commands.lua:17>
	[C]: in function 'nvim_exec2'
	vim/_editor.lua: in function 'cmd'
	...atharte/.config/nvim/lua/plugins/ui/nvimtree/actions.lua:44: in function <...atharte/.config/nvim/lua/plugins/ui/nvimtree/actions.lua:16>
stack traceback:
	[C]: in function 'nvim_exec2'
	vim/_editor.lua: in function 'cmd'
	...atharte/.config/nvim/lua/plugins/ui/nvimtree/actions.lua:44: in function <...atharte/.config/nvim/lua/plugins/ui/nvimtree/actions.lua:16>

```
</details>

I just had some spare time to debug the issue and I decided to _check_ which was the reason, and of course, it was plugin incompatibility: I've been also using [hbac.nvim](https://github.com/axkirillov/hbac.nvim) which automatically deletes _unpinned buffers_ (freeing memory I suppose) after a certain amount of buffers were opend

Not really an issue nor a bug report, its no one's fault they fundamentally conflicts with each other, just reporting this in case someone stubles across this plugin and the same issue arises.

EDIT: I suppose I should also suggest to configure vuffers as a non-lazy plugin, probably load them as late as possible in the plugin chain ([spec.priority](https://lazy.folke.io/spec#spec-loading)).

Funnily enough, if `hbac` gets configured adequately, you can actually limit the number of entries in vuffers too (e.g. in the following configuration, vuffers has a maximum of 15 entires, list won't grow further), which 
seems like a nice feature!

Here's how I configured it

```lua
return {
  'axkirillov/hbac.nvim',
    name = 'util.hbac',
    main = 'hbac',
    lazy = false,

    opts = {
      autoclose = true,
      threshold = 15,
      close_command = function(bufnr)
        local path            = vim.api.nvim_buf_get_name(bufnr)
        local status, vuffers  = pcall(require, 'vuffers.buffers.buffers')
        if status then
          vuffers.remove_buffer({ path = path }) end
        vim.api.nvim_buf_delete(bufnr, { })
      end
    },
}
```

[vuffers.buffers.buffers.remove_buffer](https://github.com/Hajime-Suzuki/vuffers.nvim/blob/main/lua/vuffers/buffers/buffers.lua#L141) is documented as it would need a bufnr, but it actually requires a path instead

Well... I don't think you meant them as an API, just pointing out :laughing:  

Still can't express enough how much I am liking this plugin so far, thanks again! (feel free to close this _issue_)

**Comment by Hajime-Suzuki** [comment]
Thank you for letting me know! If I understand correctly, you don't have the error anymore because of your setup?

Vuffefers has an autocommand on buffer closed ([here](https://github.com/Hajime-Suzuki/vuffers.nvim/blob/32c1517d5a57c0e174888545ac423e11d352f512/lua/vuffers/auto-commands.lua#L56-L82)). So I'm not really sure why the autocommand was not triggered. I need some time to check the behavior of the plugin.

I'll fix the function signature that you pointed out. Indeed it's `path` not `bufnr` 🙏

**Comment by DrKGD** [comment]
After having posted the issue, I got back at testing it (by removing the additional configuration to hbac) and... I'd say the issue was due to lazy loading vuffers.nvim (e.g. by displaying the vuffers list), so not really an issue with the plugin.
Of course, you always find the solution a little too late 🙃 I am sorry I had wasted your time.

So, a little recap, the easy solution is **not** lazy load vuffers.nvim, thus letting vuffers correctly populate the list of buffers on BufEnter, so that when hbac.nvim tries deleting a buffer, vuffers.nvim doesn't get confused over it. 

You may close the issue, not a problem "anymore", it really was due to misconfiguration, more like a heads up kind of thing for _other_ users which may be interested in both plugins.

Also I was pointing out at the integrability with hbac.nvim, you may set an upper bound to the numbers of total loaded buffers in neovim and that would automatically reflect in vuffers.nvim, kinda neat!

**Comment by Hajime-Suzuki** [comment]
Nice that you don't have the issue anymore! 👍 
When you run `setup` all the opened buffers are stored in the internal buffer lists. So I'm still not fully sure why removing lazy loading solved the problem. Anyway I will take a look sometime.

Meanwhile, I'll add the description to the readme that I suggest not to lazy load buffers.

By the way, hbac.nvim looks very nice! I'll try it.

**Comment by Hajime-Suzuki** [comment]
I tried hbac.nvim, and I got the issue indeed.
The example configuration of the plugin has:

```
  close_command = function(bufnr)
    vim.api.nvim_buf_delete(bufnr, {})
  end,
```

However, when `vim.api.nvim_buf_delete` is called I don't receive the `BufDelete` event. I tried to set the autocommand at the beginning of `.config/nvim/init.lua`, but it didn't work. Vuffers rely on the BufDelete event to maintain the internal buffer list, so missing the delete event breaks the functionality.

Since there is a workaround, I will close the issue for now. I'll take a look sometime later again.
